### PR TITLE
feat: support cycles in pathsToRoot

### PR DESCRIPTION
**Pull request by admons**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/dep-graph/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
Add support for cycles graphs in `pkgPathsToRoot` and `countPathsToRoot` functions

### CODE REVIEW

File: test/core/create-from-json.test.ts [hunk 1]
```diff
@@ -390,12 +390,6 @@ test('fromJSON with a cycle', () => {
     { name: 'bar', version: '3' },
     { name: 'baz', version: '4' },
   ]);
-
```
Comment (admons) [inline test/core/create-from-json.test.ts:393]: pkgPathsToRoot will not throw exceptions in cycles graphs anymore

**Comment by darscan** [comment]
Did you look at the work that was done here? https://github.com/snyk/dep-graph/pull/48

File: src/core/dep-graph.ts [hunk 2]
```diff
@@ -327,17 +306,26 @@ class DepGraphImpl implements types.DepGraphInternal {
     return node;
   }
 
-  private pathsFromNodeToRoot(nodeId: string): types.PkgInfo[][] {
+  private pathsFromNodeToRoot(
+    nodeId: string,
+    ancestors: string[] = [],
+  ): types.PkgInfo[][] {
+    if (ancestors.includes(nodeId)) {
+      return [];
```
Comment (darscan) [inline src/core/dep-graph.ts:314]: This feels wrong

File: src/core/dep-graph.ts [hunk 3]
```diff
@@ -327,17 +306,26 @@ class DepGraphImpl implements types.DepGraphInternal {
     return node;
   }
 
-  private pathsFromNodeToRoot(nodeId: string): types.PkgInfo[][] {
+  private pathsFromNodeToRoot(
+    nodeId: string,
+    ancestors: string[] = [],
```
Comment (darscan) [inline src/core/dep-graph.ts:311]: Is there a reason you chose an Array instead of a Set?

File: src/core/dep-graph.ts (hunk 2 above)
Comment (admons) [reply src/core/dep-graph.ts:312]: I'll move this condition into another place

File: src/core/dep-graph.ts [hunk 4]
```diff
@@ -327,26 +304,47 @@ class DepGraphImpl implements types.DepGraphInternal {
...
-      const out = this.pathsFromNodeToRoot(id).map((path) => {
-        return [this.getNodePkg(nodeId)].concat(path);
-      });
-      for (const path of out) {
-        allPaths.push(path);
-      }
-    });
+    for (const id of parentNodesIds) {
+      allPaths.push(
+        ...this.pathsFromNodeToRoot(id, ancestors.concat(nodeId)).map((path) =>
```
Comment (admons) [inline src/core/dep-graph.ts:330]: Replace with `for of` and `push` instead of spread operator

**Review by darscan — APPROVED** [review]
Approving, but please fix traversal condition and array spread as discussed

File: src/core/dep-graph.ts (hunk 3 above)
Comment (admons) [reply src/core/dep-graph.ts:309]: Consistency as we talked

**Pull request merged**